### ISSUE TITLE
Very first program

### ISSUE THREAD

**Issue by Yatekii** [warmcat/libwebsockets#892]
Hi!

I have found some sample code and since I was unable to find a basic example from the "official docs" I tried to make it work.

```
#include <math.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <libwebsockets.h>

#define PI 3.14159265

static int callback_http(struct lws_context * this,
                         struct lws *wsi,
                         enum lws_callback_reasons reason, void *user,
                         void *in, size_t len)
{
    return 0;
}

static int callback_dumb_increment(struct lws_context * this,
                                   struct lws *wsi,
                                   enum lws_callback_reasons reason,
                                   void *user, void *in, size_t len)
{
    printf("iekekekek");
    switch (reason) {
        case LWS_CALLBACK_ESTABLISHED: // just log message that someone is connecting
            printf("connection established\n");
            break;
        case LWS_CALLBACK_RECEIVE: { // the funny part
            // create a buffer to hold our response
            // it has to have some pre and post padding. You don't need to care
            // what comes there, libwebsockets will do everything for you. For more info see
            // http://git.warmcat.com/cgi-bin/cgit/libwebsockets/tree/lib/libwebsockets.h#n597
            unsigned char *buf = (unsigned char*) malloc(LWS_SEND_BUFFER_PRE_PADDING + len +
                                                         LWS_SEND_BUFFER_POST_PADDING);
           
            int i;
           
            // pointer to `void *in` holds the incomming request
            // we're just going to put it in reverse order and put it in `buf` with
            // correct offset. `len` holds length of the request.
            for (i=0; i < len; i++) {
                buf[LWS_SEND_BUFFER_PRE_PADDING + (len - 1) - i ] = ((char *) in)[i];
            }
           
            // log what we recieved and what we're going to send as a response.
            // that disco syntax `%.*s` is used to print just a part of our buffer
            // http://stackoverflow.com/questions/5189071/print-part-of-char-array
            printf("received data: %s, replying: %.*s\n", (char *) in, (int) len,
                 buf + LWS_SEND_BUFFER_PRE_PADDING);
           
            // send response
            // just notice that we have to tell where exactly our response starts. That's
            // why there's `buf[LWS_SEND_BUFFER_PRE_PADDING]` and how long it is.
            // we know that our response has the same length as request because
            // it's the same message in reverse order.
            lws_write(wsi, &buf[LWS_SEND_BUFFER_PRE_PADDING], len, LWS_WRITE_TEXT);
           
            // release memory back into the wild
            free(buf);
            break;
        }
        default:
            break;
    }
   
   
    return 0;
}

static struct lws_protocols protocols[] = {
    /* first protocol must always be HTTP handler */
    {
        "http-only",   // name
        (lws_callback_function*)callback_http, // callback
        0              // per_session_data_size
    },
    {
        "dumb-increment-protocol", // protocol name - very important!
        (lws_callback_function*)callback_dumb_increment,   // callback
        0                          // we don't use any per session data

    },
    {
        NULL, NULL, 0   /* End of list */
    }
};
 
int main(void) {
    // server url will be http://localhost:9000
    struct lws_context *context;
    struct lws_context_creation_info info;
    memset(&info, 0, sizeof info);
    info.port = 50090;

    // create libwebsocket context representing this server
    context = lws_create_context(&info);
   
    if (context == NULL) {
        fprintf(stderr, "libwebsocket init failed\n");
        return -1;
    }
   
    printf("starting server...\n");
   
    // infinite loop, to end this server send SIGTERM. (CTRL+C)
    while (1) {
        lws_service(context, 50);
        // lws_service will process all waiting events with their
        // callback functions and then wait 50 ms.
        // (this is a single threaded webserver and this will keep our server
        // from generating load while there are not requests to process)
    }
   
    lws_context_destroy(context);
   
    return 0;
}
```

I can compile it and run the server which gives the following output:

```
[2017/05/10 15:28:15:3007] NOTICE: Initial logging level 7
[2017/05/10 15:28:15:3007] NOTICE: Libwebsockets version: 2.2.0 brew@ElCapitan.local-
[2017/05/10 15:28:15:3007] NOTICE: IPV6 not compiled in
[2017/05/10 15:28:15:3007] NOTICE: libev support not compiled in
[2017/05/10 15:28:15:3007] NOTICE: libuv support not compiled in
[2017/05/10 15:28:15:3008] NOTICE:  Threads: 1 each 4864 fds
[2017/05/10 15:28:15:3008] NOTICE:  mem: platform fd map: 38912 bytes
[2017/05/10 15:28:15:3009] NOTICE:  Compiled with OpenSSL support
[2017/05/10 15:28:15:3009] NOTICE:  SSL disabled: no LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT
[2017/05/10 15:28:15:3009] NOTICE: Creating Vhost 'default' port 50090, 1 protocols, IPv6 off
[2017/05/10 15:28:15:3026] NOTICE:  Listening on port 50090
[2017/05/10 15:28:15:3026] NOTICE:  mem: per-conn:          568 bytes + protocol rx buf
[2017/05/10 15:28:15:3027] NOTICE:  canonical_hostname = noahs-mbp.edu.ds.fhnw.ch
[2017/05/10 15:28:15:3027] WARN: setgid: Operation not permitted
[2017/05/10 15:28:15:3030] WARN: setuid: Operation not permitted
starting server...
```

Sounds ok to me.

I have this js demo-testcode:

```
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js"></script>
        <script type="text/javascript">
            $(function() {
                window.WebSocket = window.WebSocket || window.MozWebSocket;
                var websocket = new WebSocket('ws://localhost:50090', 'dumb-increment-protocol'); //10.84.130.54
                websocket.onopen = function () {
                    $('h1').css('color', 'green');
                };
                websocket.onerror = function () {
                    $('h1').css('color', 'red');
                };
                websocket.onmessage = function (message) {
                    console.log(message.data);
                    $('div').append($('<p>', { text: message.data }));
                };
                
                $('button').click(function(e) {
                    e.preventDefault();
                    websocket.send($('input').val());
                    $('input').val('');
                });
            });
        </script>
        </head>
    <body>
        <h1>WebSockets test</h1>
        <form>
            <input type="text" />
            <button>Send</button>
        </form>
        <div></div>
    </body>
</html>
```

The socket can open properly apparently, but I cannot send any data ...
The server does not say anything whatsoever.

All I try to do is open a basic connection where I can push binary arrays and some JSON through ...
This was like 20lines in python and worked out of the box, so I am sure it can't be too hard in c either ...
Can anybody help me?

Thanks
Yatekii

**Comment by lws-team** [comment]
You can help yourself... when you build lws it builds its own test apps by default.  So you can, eg, run `libwebsockets-test-server` and visit `http://localhost:7681`.  The sources are in ./test-server.

The code you have pasted above has problems, with lws you must request a callback when the connection is writable, and you can only use lws_write() from that callback.  Allowing for that, looking at that code, it will only send something in response to receiving something.

**Comment by Yatekii** [comment]
Ok, I will have a look at the test-server :) Hopefully my dumb mind will understand ... And yeah I know it only reacts to sent texts, but when I send a text it doesn't do anything :)

Thanks for the quick answer!

**Comment by lws-team** [comment]
NP... it's true the test server grew until it's not a simple example.  But it has the big advantage you can confirm all the major features you might need works immediately, if you copy the related pieces out from ./test-server and can build it standalone, and that works, you can carefully cut it down checking the remaining parts you want still work at each step (git is your friend here).

**Comment by GourAbhishek** [comment]
Did the callback prototype changed ? I could see new argument 'struct lws_context * this' in callback ?

**Comment by lws-team** [comment]
Don't try to use random screenscrapes from the net.

Lws has a pile of minimal examples now (v3.0+)

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples

Look through those
